### PR TITLE
Optimizations for TypeName#toString() and CodeWriter#emit(String)

### PR DESCRIPTION
**Pull request by gfx**
This PR includes optimizations for bottlenecks: `TypeName#toString()` and `CodeWriter#emit(String)`
Especially `TypeName#toString()` is related to `#equals()` and `#hashCode()`, meaning `TypeName` becomes suitable for `HashMap` keys.

## Benchmark

Given the following code:

```java
package com.squareup.javapoet;
import org.junit.Before;
import org.junit.Test;
import javax.lang.model.element.Modifier;
import java.util.Collections;
import static com.google.common.truth.Truth.assertThat;
public class PerformanceTest {
    JavaFile javaFile;
    @Before
    public void setUp() throws Exception {
        ClassName hoverboard = ClassName.get("com.mattel", "Hoverboard");
        ClassName namedBoards = ClassName.get("com.mattel", "Hoverboard", "Boards");
        ClassName list = ClassName.get("java.util", "List");
        ClassName arrayList = ClassName.get("java.util", "ArrayList");
        TypeName listOfHoverboards = ParameterizedTypeName.get(list, hoverboard);
        MethodSpec beyond = MethodSpec.methodBuilder("beyond")
                .returns(listOfHoverboards)
                .addStatement("$T result = new $T<>()", listOfHoverboards, arrayList)
                .addStatement("result.add($T.createNimbus(2000))", hoverboard)
                .addStatement("result.add($T.createNimbus(\"2001\"))", hoverboard)
                .addStatement("result.add($T.createNimbus($T.THUNDERBOLT))", hoverboard, namedBoards)
                .addStatement("$T.sort(result)", Collections.class)
                .addStatement("return result.isEmpty() $T.emptyList() : result", Collections.class)
                .build();

        MethodSpec m = MethodSpec.methodBuilder("z")
                .returns(String.class)
                .addStatement("return $S", "Hello, world!")
                .addModifiers(Modifier.PUBLIC).build();
        TypeSpec hello = TypeSpec.classBuilder("HelloWorld")
                .addMethod(beyond)
                .addMethod(MethodSpec.methodBuilder("a")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("b")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("c")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("d")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("d")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("f")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("g")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("h")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("i")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .addMethod(MethodSpec.methodBuilder("j")
                        .returns(String.class)
                        .addStatement("return $S", "Hello, world!")
                        .addModifiers(Modifier.PUBLIC).build())
                .build();
        javaFile = JavaFile.builder("com.example.helloworld", hello)
                .addStaticImport(hoverboard, "createNimbus")
                .addStaticImport(namedBoards, "*")
                .addStaticImport(Collections.class, "*")
                .build();
    }

    @Test
    public void importStaticReadmeExample() {
        long t0 = System.currentTimeMillis();
        for(int i = 0; i < 10000; i++) {
            @SuppressWarnings("unused") String s = javaFile.toString();
        }
        System.out.println(System.currentTimeMillis() - t0);

        t0 = System.currentTimeMillis();
        for(int i = 0; i < 10000; i++) {
            @SuppressWarnings("unused") String s = javaFile.toString();
        }
        System.out.println(System.currentTimeMillis() - t0);

        t0 = System.currentTimeMillis();
        for(int i = 0; i < 10000; i++) {
            @SuppressWarnings("unused") String s = javaFile.toString();
        }
        System.out.println(System.currentTimeMillis() - t0);
    }
}
```

In my environment, Macbook Air 2014, the result is something like this (unit: milliseconds):

`HEAD`:
3565
1264
1194

(1) TypeName#toString() cache 2660602 :
3424
1013
998

(2) CodeWriter#emit(String) cdc51fd :
3339
1003
1012

(1) + (2):
3009
1245
740

I think it is worth merging. What do you think of it?

### CODE REVIEW

**Comment by ronshapiro** [comment]
A couple notes on the benchmarking: `javac` might be optimizing some of those loops since they don't produce anything. The best way to account for that is to have an `int result` that you continue adding the hash code of `javaFile.toString()`, then printing that at the end. Worth looking into [Caliper](https://www.youtube.com/watch?v=Rbx0HUCnF24) if you want to do more micro bencharks.

**Comment by sormuras** [comment]
Or [JMH](http://openjdk.java.net/projects/code-tools/jmh/)

**Comment by gfx** [comment]
Sure. Changed the benchmark code to:

```java
    @Test
    public void importStaticReadmeExample() {
        long x = 0;
        long t0 = System.currentTimeMillis();
        for(int i = 0; i < 10000; i++) {
            x += javaFile.toString().hashCode();
        }
        System.out.println(System.currentTimeMillis() - t0);

        t0 = System.currentTimeMillis();
        for(int i = 0; i < 10000; i++) {
            x += javaFile.toString().hashCode();
        }
        System.out.println(System.currentTimeMillis() - t0);

        t0 = System.currentTimeMillis();
        for(int i = 0; i < 10000; i++) {
            x += javaFile.toString().hashCode();
        }
        System.out.println(System.currentTimeMillis() - t0);

        assertNotEquals(x, 0L);
    }
}
```

and confirm it produces almost the same scores as before.

I'll use JMH for the next time.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 1]
```diff
@@ -202,6 +202,10 @@ public void emitTypeVariables(List<TypeVariableName> typeVariables) throws IOExc
     emit(">");
   }
 
+  public CodeWriter emit(String s) throws IOException {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Unused?

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Ahh, this is the 0-element case on varargs.

**Comment by swankjesse** [comment]
Thanks! I manually tweaked this to avoid a theoretical concurrency problem. Basically you want to read the field only once, and you can do this by creating a local variable.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 1 above)
Comment (gfx) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Right. In order to omit parsing `format` string.

**Comment by gfx** [comment]
Thanks!

**Pull request closed without merging**